### ISSUE TITLE
Tokyo Disneyland - Fantasy Tour hanging/crash

### ISSUE THREAD

**Issue by medigi** [TASVideos/BizHawk#2482]
Tokyo Disneyland - Fantasy Tour (Japan)
SHA1:E413B0AB506CB12B68439CCC0A72E4664C112378
MD5:6DB7AA2399E9B6723E41E47C73379216

Crashing in Minnie's mini-game "Minnie's House" for unknown for me reason. Tested in cores: Gambatte, SameBoy, GBHawk (Bizhawk 2.5.2)
[Tokyo Disneyland - Fantasy Tour states.zip](https://github.com/TASVideos/BizHawk/files/5553968/Tokyo.Disneyland.-.Fantasy.Tour.states.zip)

For reproduce, load savestate in any of these cores and press "A" button.

**Comment by nattthebear** [comment]
Crashes in three completely unrelated cores?  Hmm

**Comment by medigi** [comment]
Yes, except BSNES core

**Comment by nitro2k01** [comment]
This seems to happen because unmapped SRAM reads `$FF` which is generally correct because there are pull-up resistors, but on hardware it's a bit more complicated than that. Because of bus capacitance, the last value that existed on the bus (in this case the opcode for the read) remains for a short period when reading from non-existent or disabled SRAM. Note that this is an analog effect and the exact behavior may vary depending on the model and revision of Gameboy as well as the cartridge being used. It's also not always deterministic.

Observations from my tests on two Gameboys so far. 
- On a DMG, it consistently reads back the value of the read opcode. (In my test this was `ld A,[DE]` and read back `$1F`)
- On a GBA SP, the value reads back using the same code varied and could be any one of `$1A,$5A,$DA,$FA`. This indicates that the voltage level on the bus is right at the logic level threshold and crosses it on some cycles but not on other cycles for the various data lines.

As for the bug in the game in question, the gory details are as follows: The game is using a table to look up the VRAM locations for the items in the game. Unfortunately, the code at `08:4BAB` has a bug and iterates 17 times instead of 16. (The number of items in the minigame is 16.) 

At `08:4B79` this leads to an out of bounds read from the table starting at `$4FFB`. The out of bound read reads a pointer from `$501B` which is the value `$A698`, which points to SRAM. This is the first value of the next table and is actually supposed to represent `$98A6` in the opposite endianess, which is a target address in VRAM. 

The effect is that `$A698` is read and treated as a target pointer for a write later down the line. It's read with the instruction `ld A,[BC]` which has opcode `$0A`.

- On hardware, this typically means a value like `$0A0A` which is a harmless write to the MBC.
- On most emulators, this produces a write to `$FFFF` in the end, which happens to be the interrupt enable register. Because of the data written, this may disable the VBlank interrupt and make the game freeze.

You can prove that this really is what is causing the freeze using BGB. Set an access breakpoint to reads from SRAM. (`A000-BFFF`) Trigger the bug in the minigame. When it halts, step over the read, edit the `A` register so it is `0A`. Do this again for the second read. Observe that the game does not crash.

**Comment by CasualPokePlayer** [comment]
To elaborate further on this "non-deterministic" behavior, the results can be affected by temperature, dirt on the pins, etc. And you can see with the GBA SP you can just already see non-deterministic results (minor note, the GB Player has results like the DMG nitro tested) And then you can consider that for some systems, the FF read could be correct and you might actually find for some systems the game would actually crash like emulators do.

Also for why BSNES doesn't crash, this code is likely to blame:
> uint8 Cartridge::ram_read(unsigned 0addr) {
if(ramsize == 0) return 0x00;
0x00 reads are just impossible (0x00 is nop, and pullups only turn 0s into 1s, not vice versa). So it doesn't crash on BSNES due to an inaccuracy, although other emus it crashes due to them being closer to being truly accurate.

**Comment by alyosha-tas** [comment]
Neat, thanks for the breakdown. Is this a general thing or only MBC1?

**Comment by CasualPokePlayer** [comment]
General thing, lack of SRAM/locked SRAM have fun:tm: open bus results.

**Comment by alyosha-tas** [comment]
I implemented basic open bus behaviour for bad SRAM accesses for mbc1 in GBHawk. I'm not going to spend time on analog effects here.

**Comment by CasualPokePlayer** [comment]
The behaviour definitely affects other MBCs, my tests with the GB Player have been with an MBC3 flashcart, and previous tests have been done using ACE on Pokemon Gold, which uses MBC3. Also definitely doesn't need an MBC, no SRAM is essentially the same as locked SRAM for these purposes.

The fix is also just inaccurate, and goes under the assumption everything uses the same memory bus, when in reality they don't. For the GB Player at least, ROM and SRAM share their bus, while VRAM/WRAM/OAM/HRAM do not share that bus. This matters since the read does not necessarily have to be executed in ROM, it could be executed from the 4 memory chunks I laid out that do not share their bus with SRAM. If the read is executed from one of those 4 locations, nothing is going to be pushed into the ROM/SRAM bus at all (the opcode value will go into a different bus), so you'll end up reading 0xFF. This is only coming from my tests on the GB Player, bus arrangements can vary based on revision (but I'm pretty sure in no model everything would share same bus). 

This also assumes every opcode will produce this behavior for all reads, which is just incorrect (and generally needs much more testing). For every test to my knowledge, only opcodes without operands have been used. For an opcode with an operand, let's take ld a,[xxyy]. Would it return FA? Or would it return A0-BF? The CB prefix also applies with this. And there is the case with pop/ret. pop/ret do 2 separate 8-bit reads. From my testing on the GB Player, the first read returns the last opcode executed, while the second read will just return FF.

There is also a fast elephant in the room, being double speed mode. While this is more for the GBC/GBA, it hasn't been tested as far as I know, although it likely would have an effect on this behavior.

Even ignoring the freezer and sticking to the practically deterministic results, there's still a lot that needs testing regarding this behavior.

**Comment by alyosha-tas** [comment]
Ok I updated the emulation of this to only effect the cart busses. It still won't do the 2 consecutive reads thing correctly, but it's good enough for now.

It's pretty expandable and doesn't really slow things down so it can be improved later if necessary. But I'll probably only do so if some specific game or TAS is actually broken.

**Comment by CasualPokePlayer** [comment]
Seems like my suspicions were right about operands, ld a,[xxyy] does indeed take in the last operand (ie, for reading to $A000, $A0 would be returned). Goes to show that knowledge is incomplete with this lol

**Comment by alyosha-tas** [comment]
default mapper (no MBC) does return open bus if no SRAM is present. Open bus behaviour is emulated on no MBC, MBC1 and MBC3 currently.

**Comment by nitro2k01** [comment]
> EDIT: Well, the MBC2 is a potential odd case, since the upper 4 bits of RAM are open bus, but the lower 4 bits have an actual value, not sure what the upper 4 bits should have (probably following current ff approach just make it OR 0xF0? Don't know how the last opcode thing would work out?)

When SRAM is disabled it should nominally work the same as anything else because the MBC is not driving the data bus. In practice, there may some small timing difference between the upper and lower nibble because the high impedance MBC connection may still make the data bus settle to `FF` slightly faster or slower.

Nintendo could however have decided to implement the SRAM weirdly due to the integration of SRAM into the MBC, including:

- Reads actually still work as normal and only writes are disabled.
- The lower nibble is driven, but to `0` or `F` and not the value stored in RAM.

When SRAM is enabled, the upper nibble should exhibit open bus, and show the last value that was on the bus, for the upper nibble only. However, I would expect the impact of getting this wrong is low in terms of games malfunctioning, since it was well documented by Nintendo that the upper nibble should be discarded, and most games probably did do so in practice. Because of the small size of the SRAM, it's also likely that most games only access it one place to copy it in or out of SRAM, making bugs that rely on the upper nibble less likely.

Over all, more research needed!

**Comment by medigi** [comment]
Just tested this game in the new implented bsnes version 115 (as a new core) = doesn't crashing. However, if possible, update SameBoy core so i can test the full entire game, and if no crash - issue can be closed as fixed

**Comment by alyosha-tas** [comment]
I just assumed writes would effect the bus as well. Once an actual test is out there verifying otherwise, I can easily fix it.

**Comment by medigi** [comment]
Tested this game on SameBoy 0.14.3 and confirming. This is probably one of problematic gameboy games to emulate

**Comment by CasualPokePlayer** [comment]
> Tested this game on SameBoy 0.14.3 and confirming a crash on that mini-game. This is probably one of problematic gameboy games to emulate
No, the entire problem is disabled SRAM read behavior. It's rather just due to a bug in this game, and the crashing behavior seem on emulators could very well happen on certain consoles too, as disabled SRAM reads are (more technically) non-deterministic (more practically, they are deterministic with certain consoles).
